### ISSUE TITLE
False negative for used-before-assignment and unused-argument when used in inner function without nonlocal

### ISSUE THREAD

**Issue by LLyaudet** [pylint-dev/pylint#9512]
### Bug description

```python
"""A module doctring"""

def test_closure_provider(y):
    """A doc string"""
    def test_closure_result(z):
        """Another doc string"""
        while y > 0:
            z *= 2
            y -= 1
        return z
    return test_closure_result
```

### Command used

```shell
pylint pylint_closure.py
```

### Pylint output

```python
************* Module pylint_closure
pylint_closure.py:7:14: E0601: Using variable 'y' before assignment (used-before-assignment)
pylint_closure.py:3:26: W0613: Unused argument 'y' (unused-argument)
```

### Expected behavior

no error

### Pylint version

```shell
$ pylint --version
pylint 3.1.0
astroid 3.1.0
Python 3.11.6 (main, Oct  8 2023, 05:06:43) [GCC 13.2.0]
```

### OS / Environment

Ubuntu 23.10

**Comment by LLyaudet** [comment]
Hello :)
I really tried to get a minimal example, inner function is required and while loop also.
I hope it helps.
Best regards,
    Laurent Lyaudet

**Comment by LLyaudet** [comment]
My bad, I thought that Python dealed with closures like JS. I added nonlocal keyword and the error is gone.
But I had other arguments without the error so it is not a False positive it is a False negative:

I should have the error and the warning with the following code without the while loop and without nonlocal.
And I have None.
```
"""A module doctring"""

def test_closure_provider(y):
    """A doc string"""

    def test_closure_result(z):
        """Another doc string"""
        return z + y

    return test_closure_result
```

**Comment by mbyrnepr2** [comment]
Hi @LLyaudet :)

My understanding is that in the first example the assignment `y -= 1` is the problematic part: this assignment makes `y` refer to something that is local to the `test_closure_result` function and stops referring to the parameter of test_closure_provider.
```python
"""example1.py"""
def test_closure_provider(y):
    """A doc string"""
    def test_closure_result(z):
        """Another doc string"""
        while y > 0:
            z *= 2
            y -= 1
        return z
    return test_closure_result

one = test_closure_provider(1) 
print(one(1))
```

python example1.py:
```python
  File "/Users/markbyrne/programming/y.py", line 7, in test_closure_result
    while y > 0:
          ^
UnboundLocalError: cannot access local variable 'y' where it is not associated with a value
```


The above issue is not realised in the second example because the inner function is simply reading `y`:
```python
"""example2.py"""

def test_closure_provider(y):
    """A doc string"""

    def test_closure_result(z):
        """Another doc string"""
        return z + y

    return test_closure_result

one = test_closure_provider(1)
print(one(1))
```

python example2.py:
`2`

SO post similar issue: https://stackoverflow.com/questions/29639780/unboundlocalerror-local-variable-referenced-before-assignment-in-python-closure

**Comment by LLyaudet** [comment]
Thanks @mbyrnepr2  :)
I'm not a specialist of nonlocal keyword.
Now I understand that it is sometimes needed and sometimes not,
because Python has a "best effort" approach until both behaviours may be possible and then the use of nonlocal becomes significant :)
